test/unit/org/apache/cassandra/metrics/TableMetricsTest.java

```diff
@@ -363,6 +363,63 @@ public void testViewMetricsCleanupOnDrop()
+        assertEquals(0L, cfs.metric.estimatedPartitionCountInSSTablesCached.getValue().longValue());
+        long startTime = System.currentTimeMillis();
+
+        int partitionCount = 10;
+        int numRows = 100;
+
+        for (int i = 0; i < numRows; i++)
+            session.execute(String.format("INSERT INTO %s.%s (id, val1, val2) VALUES (%d, '%s', '%s')", KEYSPACE, TABLE, i % partitionCount, "val" + i, "val" + i));
+
+        assertEquals(partitionCount, cfs.metric.estimatedPartitionCount.getValue().longValue());
+        cfs.forceBlockingFlush(ColumnFamilyStore.FlushReason.UNIT_TESTS);
+        assertEquals(partitionCount, cfs.metric.estimatedPartitionCount.getValue().longValue());
+
+        long estimatedPartitionCountInSSTables = cfs.metric.estimatedPartitionCountInSSTablesCached.getValue().longValue();
+        long elapsedTime = System.currentTimeMillis() - startTime;
+        // the caching time is one second; avoid flakiness by only checking if a long time has not passed
+        if (elapsedTime < 980)
+            assertEquals(0, estimatedPartitionCountInSSTables);
+
+        for (int i = 0; i < numRows; i++)
+            session.execute(String.format("INSERT INTO %s.%s (id, val1, val2) VALUES (%d, '%s', '%s')", KEYSPACE, TABLE, i % partitionCount, "val" + i, "val" + i));
+
+        estimatedPartitionCountInSSTables = cfs.metric.estimatedPartitionCountInSSTablesCached.getValue().longValue();
+        elapsedTime = System.currentTimeMillis() - startTime;
+        if (elapsedTime < 980)
+            assertEquals(0, estimatedPartitionCountInSSTables);
+        else if (elapsedTime >= 1020)
+            assertEquals(partitionCount, estimatedPartitionCountInSSTables);
+
+        // The answer below is incorrect but what the metric currently returns.
+        assertEquals(partitionCount * 2, cfs.metric.estimatedPartitionCount.getValue().longValue());
+        cfs.forceBlockingFlush(ColumnFamilyStore.FlushReason.UNIT_TESTS);
+        // Recalculation for the new sstable set will correct it.
+        assertEquals(partitionCount, cfs.metric.estimatedPartitionCount.getValue().longValue());
+
+        // The cached estimatedPartitionCountInSSTables lags one second, check that.
+        estimatedPartitionCountInSSTables = cfs.metric.estimatedPartitionCountInSSTablesCached.getValue().longValue();
+        elapsedTime = System.currentTimeMillis() - startTime;
+        if (elapsedTime < 980)
+            assertEquals(0, estimatedPartitionCountInSSTables);
+        else if (elapsedTime >= 1020)
+            assertEquals(partitionCount, estimatedPartitionCountInSSTables);
+
+        // Wait to let it update.
+        if (elapsedTime < 1020)
+            Thread.sleep(1200 - elapsedTime);
+
+        // It must now report the correct value.
+        estimatedPartitionCountInSSTables = cfs.metric.estimatedPartitionCountInSSTablesCached.getValue().longValue();
+        assertEquals(partitionCount, estimatedPartitionCountInSSTables);
```

Comment:
question:
Wouldn't it be simpler to just assert `assertEquals(partitionCount, estimatedPartitionCountInSSTables)` with some backoff? It seems to me that's the ultimate goal we are trying to achieve.

Reply:
Could you point to a sample or suggest the relevant code change?

(I haven't used a waiting check before and don't know where to look for an example.)

Reply:
Example: https://github.com/datastax/cassandra/blob/main/test/unit/org/apache/cassandra/io/sstable/SSTableIdTest.java#L271.
```suggestion
        // Assert that the metric will return a correct value after at least a second passes
        await().atMost(2, TimeUnit.SECONDS)
                .untilAsserted(() -> {
                    long value = cfs.metric.estimatedPartitionCountInSSTablesCached.getValue();
                    assertEquals(partitionCount, value);            
                });
```

Reply:
It seems the compilation started failing after applying this suggestion due to the lack of an import (sorry I didn't notice earlier).
```
import static org.awaitility.Awaitility.await;
```
should resolve it.